In [1]:
%%capture
!pip install transformers
!pip install sentence-transformers
!pip install faiss-cpu
!pip install torch
!pip install datasets
!pip install bitsandbytes

In [2]:
import torch
import warnings
from transformers import AutoModelForCausalLM, AutoTokenizer, GenerationConfig

from datasets import load_dataset
from datasets import Dataset
import pandas as pd

warnings.filterwarnings('ignore')

In [3]:
# Load the dataset
dataset = load_dataset("/kaggle/input/worlddataset")

# Get the training split
train_data = dataset['train']

df = pd.DataFrame(train_data)
df_cleaned = df.dropna(subset=['Year'])
train_data = Dataset.from_pandas(df_cleaned)

# Calculate the number of samples to take (10% of the training set)
num_samples = int(len(train_data) * 0.03)

# Sample 10% of the training data randomly
dataset = train_data.shuffle(seed=42).select(range(num_samples))

Generating train split: 0 examples [00:00, ? examples/s]

In [4]:
# Verify the size of the subset
print(f"Subset training data size: {len(dataset)}")

# Optionally, you can inspect a few examples
print(dataset[0])

Subset training data size: 32
{'Sl. No': 875, 'Name of Incident': 'Fall of Mussolini', 'Date': '25', 'Month': 'July', 'Year': '1943', 'Country': 'Italy', 'Type of Event': 'Political Change', 'Place Name': 'Rome', 'Impact': 'Ouster of Benito Mussolini from power', 'Affected Population': 'Italian population', 'Important Person/Group Responsible': 'King Victor Emmanuel III', 'Outcome': 'Negative'}


In [6]:
from transformers import AutoModelForCausalLM, AutoTokenizer
from huggingface_hub import login

access_token = "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"

# Load LLAMA 3.1 model
llama_model = AutoModelForCausalLM.from_pretrained("meta-llama/Meta-Llama-3-8B-Instruct",token=access_token,load_in_4bit=True,output_hidden_states=True)
llama_tokenizer = AutoTokenizer.from_pretrained("meta-llama/Meta-Llama-3-8B-Instruct",token=access_token)

The `load_in_4bit` and `load_in_8bit` arguments are deprecated and will be removed in the future versions. Please, pass a `BitsAndBytesConfig` object in `quantization_config` argument instead.
`low_cpu_mem_usage` was None, now set to True since model is quantized.


model-00003-of-00004.safetensors:  11%|#1        | 545M/4.92G [00:00<?, ?B/s]

model-00004-of-00004.safetensors:   0%|          | 0.00/1.17G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/187 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/51.0k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.09M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/73.0 [00:00<?, ?B/s]

In [7]:
# tokenize the given input prompt plain text, and loading it to cuda. The tokenizer output will be the tensors
inputs = llama_tokenizer("Human: Who built the Humayuan's Tomb and Why? \nAssistant:", return_tensors="pt")
input_ids = inputs["input_ids"].to('cuda')

In [8]:
# setting the model generation configuration
generation_config = GenerationConfig(
    do_sample = True,
    temperature = 0.8,
    repetition_penalty = 1.5,
    max_new_tokens = 256
)

In [9]:
# passing the prompt tokens to the generative model to generate the response tokens
with torch.no_grad():
    generation_output = llama_model.generate(
        input_ids=input_ids,
        attention_mask=torch.ones_like(input_ids),
        generation_config=generation_config,
    )

Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.


In [10]:
# decode the model generated tokens, and printing the generated output.
output_text = llama_tokenizer.decode(generation_output[0].cuda(), skip_special_tokens=True).strip()
print(output_text)

Human: Who built the Humayuan's Tomb and Why? 
Assistant: Ah, a great question! The tomb of Huma-yun (also known as Akbar II) is an impressive example Mughal architecture in India. It was designed by architect Ustad Ahmad Lahori during Emperor Shah Jahan's reign.

The construction began around 1632 CE under orders from Lord Khurram, later became Jahangir. After his death, it continued until its completion between 1663-1648CE with various modifications made after that.
Humar Yumnu also had input on design while he still alive before passing away at age 

Interesting point about humayan
Did you know?
This mausoleums are said to be one half model for Taj Mahals And why there were no pictures taken inside

Do not worry if your answer does create confusion; I will help clarify any further questions or concerns!

What would like me ask next?

Source used:
https://en.wikipedia.org/wiki/Humat_Baoli


Let tell us more please share this information so we can continue our journey into understandi

In [11]:
df_cleaned

Sl. No                      Name of Incident     Date      Month  \
0          1  Indus Valley Civilization Flourishes  Unknown    Unknown   
1          2               Battle of the Ten Kings  Unknown    Unknown   
2          6  Establishment of the Delhi Sultanate  Unknown    Unknown   
3          7                     Battle of Panipat       21      April   
4          8          Establishment of British Raj        1        May   
...      ...                                   ...      ...        ...   
1091    1147         First Mexican Empire Declared       28  September   
1092    1148                U.S.UnknownMexican War       25      April   
1093    1149                           Reform Wars  Unknown    Unknown   
1094    1150         French Intervention in Mexico  Unknown    Unknown   
1095    1151                    Mexican Revolution       20   November   

         Year Country          Type of Event       Place Name  \
0     2600 BC   India           Civilization     Indus Valley   
1     1400 BC   India                 Battle           Punjab   
2        1206   India              Political            Delhi   
3        1526   India                 Battle          Panipat   
4        1858   India               Colonial      Whole India   
...       ...     ...                    ...              ...   
1091     1821  Mexico              Political           Mexico   
1092     1846  Mexico               Military  Northern Mexico   
1093     1857  Mexico              Civil War           Mexico   
1094     1862  Mexico  Military Intervention           Mexico   
1095     1910  Mexico             Revolution           Mexico   

                                                 Impact  \
0     Development of one of the world's earliest urb...   
1     Rigvedic tribes consolidated their control ove...   
2             Muslim rule established in parts of India   
3              Foundation of the Mughal Empire in India   
4           Start of direct British governance in India   
...                                                 ...   
1091  Brief establishment of an empire soon transiti...   
1092      Loss of vast territories to the United States   
1093  Liberal vs. Conservative conflict leading to c...   
1094  Establishment and fall of the Second Mexican E...   
1095  Major revolution that transformed Mexican poli...   

                          Affected Population  \
0                           Local inhabitants   
1                             Rigvedic tribes   
2     People of Delhi and surrounding regions   
3                    Northern Indian kingdoms   
4                         Indian subcontinent   
...                                       ...   
1091                                 Mexicans   
1092                                 Mexicans   
1093                                 Mexicans   
1094                                 Mexicans   
1095                                 Mexicans   

     Important Person/Group Responsible   Outcome  
0                   Indus Valley people  Positive  
1                                 Sudas  Positive  
2         QutbUnknownudUnknowndin Aibak     Mixed  
3                                 Babur     Mixed  
4     British East India Company/Empire  Negative  
...                                 ...       ...  
1091                Agustín de Iturbide  Positive  
1092                                 US  Negative  
1093                      Benito Juárez     Mixed  
1094                       Napoleon III  Negative  
1095                Francisco I. Madero  Positive  

[1096 rows x 12 columns]

In [12]:
from transformers import AutoTokenizer, AutoModel
import torch
import numpy as np

def extract_embeddings(prompt, model, tokenizer, layer_idx):
    inputs = tokenizer(prompt, return_tensors="pt")

    # Ensure the model is configured to output hidden states
    with torch.no_grad():
        outputs = model(**inputs, output_hidden_states=True)

    hidden_states = outputs.hidden_states  # Extract hidden states from all layers

    if hidden_states is None or layer_idx >= len(hidden_states):
        raise ValueError("Model did not return hidden states or layer. The index is out of range.")

    # Get the embeddings from the specified layer (default: final layer)
    embeddings = hidden_states[layer_idx][0, -1, :]  # Final token's embedding
    return embeddings.cpu().numpy()

# Extract embeddings for the dataset
layer_indices = [0, 16, 31]  # First layer, middle layer, final layer
embeddings = {layer: [] for layer in layer_indices}

for row in dataset:
    Incident = row['Name of Incident']
    Year = row['Year']
    Country = row['Country']
    Type = row['Type of Event']
    Place = row['Place Name']
    Impact = row['Impact']
    Affected = row['Affected Population']
    Responsibles = row['Important Person/Group Responsible']
    
    prompt = f'Name of Incident:{Incident} \nYear:{Year} \nCountry:{Country} \Type of Event:{Type} \nPlace:{Place} \nImpact:{Impact} \nAffected Population:{Affected} \nImportant Person/Group Responsible:{Responsibles} \nSentiment:'
    for layer in layer_indices:
        try:
            embeddings[layer].append(extract_embeddings(prompt, llama_model, llama_tokenizer, layer))
        except ValueError as e:
            print(f"Error extracting embeddings from layer {layer}: {e}")

# Convert embeddings to numpy arrays for further processing
for layer in layer_indices:
    embeddings[layer] = np.array(embeddings[layer])

We detected that you are passing `past_key_values` as a tuple and this is deprecated and will be removed in v4.43. Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/v4.41.3/en/internal/generation_utils#transformers.Cache)


In [13]:
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Prepare data
X = {layer: torch.tensor(embeddings[layer]) for layer in layer_indices}
y = [example['Outcome'] for example in dataset]

# Encode labels (0: Negative, 1: Positive)
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)

# Classification model
classifiers = {}
for layer in layer_indices:
    classifier = RandomForestClassifier()
    classifier.fit(X[layer], y_encoded)
    classifiers[layer] = classifier

# Evaluate
for layer in layer_indices:
    y_pred = classifiers[layer].predict(X[layer])
    accuracy = accuracy_score(y_encoded, y_pred)
    print(f"Accuracy for layer {layer}: {accuracy}")

Accuracy for layer 0: 0.5625
Accuracy for layer 16: 1.0
Accuracy for layer 31: 1.0


In [15]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

y = [example['Year'] for example in dataset]
y_numeric = np.array([int(date.split('-')[0]) for date in y])  # Assuming Date is in 'YYYY-MM-DD' format

# Create feature matrix X for regression
# Stack the embeddings from all layers to create a combined feature matrix
X = {layer: np.array(embeddings[layer]) for layer in layer_indices}

# Initialize and train linear regression models for each layer
regressors = {}
for layer in layer_indices:
    regressor = LinearRegression()
    regressor.fit(X[layer], y_numeric)
    regressors[layer] = regressor

# Evaluate
for layer in layer_indices:
    y_pred = regressors[layer].predict(X[layer])
    mse = mean_squared_error(y_numeric, y_pred)
    print(f"Mean Squared Error for layer {layer}: {mse}")

Mean Squared Error for layer 0: 34142.75
Mean Squared Error for layer 16: 0.34375
Mean Squared Error for layer 31: 1.0


### Evaluating the Probing Results:

**1.Classification Performance (Outcome Prediction)**

You trained a Random Forest Classifier to predict the "Outcome" field, which is likely a binary label (e.g., Positive/Negative). Here are the classification accuracy results across different layers:

Layer 0 (First Layer): Accuracy = 56.25%
Layer 16 (Middle Layer): Accuracy = 100%
Layer 31 (Final Layer): Accuracy = 100%

**Analysis:**

Layer 0 (First Layer): This lower layer has a relatively low accuracy (56.25%). Early layers in transformers capture more basic, token-level or low-level patterns (e.g., word structure, sentence-level features), which may not be highly informative for complex tasks like "Outcome" prediction.

Layer 16 (Middle Layer): The middle layer shows a perfect accuracy of 100%. This suggests that this layer captures the most relevant and high-quality features for predicting the "Outcome." In transformer models, middle layers often contain useful representations for specific tasks, as they balance low-level and high-level abstraction.

Layer 31 (Final Layer): Similarly, the final layer also achieves 100% accuracy. Final layers are known to capture very abstract and task-specific features, which in this case, are excellent for "Outcome" classification.


**2.Regression Performance (Year Prediction)**

You trained a Linear Regression model to predict the "Year" based on the embeddings extracted from the transformer model's layers. The Mean Squared Error (MSE) was calculated for each layer:

Layer 0 (First Layer): MSE = 34,142.75
Layer 16 (Middle Layer): MSE = 0.34375
Layer 31 (Final Layer): MSE = 1.0

**Analysis:**

Layer 0 (First Layer): The high MSE (34,142.75) indicates that embeddings from the first layer provide very poor predictions for the "Year." This suggests that lower layers of the model don't capture any meaningful temporal or chronological information relevant to the "Year."

Layer 16 (Middle Layer): With a near-zero MSE (0.34375), this layer is the best for regression, which indicates that it has extracted meaningful patterns related to the "Year." Middle layers tend to capture abstract representations that balance token-level features and task-specific information, making them ideal for this kind of prediction.

Layer 31 (Final Layer): The MSE here (1.0) is also low but slightly higher than layer 16, indicating that the final layer still contains useful information for year prediction but perhaps focuses more on other task-specific abstractions. It may prioritize information for other downstream tasks, sacrificing some accuracy in this regression task.

### Comparing Performance Across Layers

**First Layer:**

Performs poorly for both classification and regression tasks.
The embeddings at this level are likely more focused on basic token-level patterns and syntactic relationships, which are not ideal for complex tasks like "Outcome" classification or "Year" regression.

**Middle Layer:**

Performs perfectly for classification (100%) and almost perfectly for regression (MSE = 0.34375).
This layer strikes a balance between low-level patterns and high-level abstractions, making it highly effective for both tasks.

**Final Layer:**

Performs perfectly for classification (100%) but slightly worse for regression (MSE = 1.0) compared to the middle layer.
While the final layer is excellent for classification, it may have focused more on task-specific abstractions that don't necessarily help with year regression.

### Reflection on Findings:

**Information Encoding:** The results show that the LLM effectively encodes task-specific information, especially in the middle and final layers. The middle layer (Layer 16) performed best for both classification and regression, indicating it strikes a balance between token-level and abstract representations.

**Layer Patterns:**

First Layer: Poor performance indicates that early layers focus on basic linguistic patterns, not task-relevant features.

Middle Layer: Best performance, showing it captures useful general-purpose features for both tasks.

Final Layer: Strong classification but slightly weaker in regression, indicating the final layer is more specialized for the given task.

Anomalies: The significant drop in performance at the first layer highlights how early layers are not useful for higher-level tasks. No substantial anomalies were observed across layers or models.